app/src/main/java/com/brandon/forzenbook/viewmodels/LoginViewModel.kt

```diff
@@ -0,0 +1,291 @@
+    private val createUserUseCase: CreateUserUseCase,
+) : ViewModel() {
+
+    private var sqlDate: Date? = null
+
+    private val _state: MutableState<LoginUiState> = mutableStateOf(LoginUiState.Idle())
+    val state: MutableState<LoginUiState>
+        get() = _state
+
+    private val _createAccountState: MutableState<CreateAccountUiState> =
+        mutableStateOf(CreateAccountUiState.Idle())
+    val createAccountState: MutableState<CreateAccountUiState>
+        get() = _createAccountState
+
+    sealed interface LoginUiState {
+
+        data class Idle(
+            val email: String = "",
+            val isEmailError: Boolean = false,
+            val isGenericError: Boolean = false,
+            val isCodeSent: Boolean = false,
+        ) : LoginUiState
+
+        data class Loading(
+            val email: String = "",
+            val isCodeSent: Boolean = false,
+        ) : LoginUiState
+
+        // TODO remove this once we have another screen FA-83
+        object LoginWithCode : LoginUiState
+
+    }
+
+    sealed interface CreateAccountUiState {
+
+        data class Idle(
+            val email: String = "",
+            val firstName: String = "",
+            val lastName: String = "",
+            val location: String = "",
+            val dateOfBirth: String = "",
+            val isEmailError: Boolean = false,
+            val isLocationError: Boolean = false,
+            val isDateError: Boolean = false,
+            val isDuplicateUser: Boolean = false,
+        ) : CreateAccountUiState
+
+        data class Loading(
+            val email: String = "",
+            val firstName: String = "",
+            val lastName: String = "",
+            val location: String = "",
+            val dateOfBirth: String = "",
+        ) : CreateAccountUiState
+
+        object Loaded : CreateAccountUiState
+    }
+
+    fun getCode(email: String) {
+        viewModelScope.launch {
+            state.value = LoginUiState.Loading(isCodeSent = false, email = email)
+            if (!isValidEmail(email)) {
+                state.value = LoginUiState.Idle(isEmailError = true, email = email)
+                cancel()
+            } else {
+                if (getLoginCodeUseCase(email)) {
+                    state.value = LoginUiState.Idle(isCodeSent = true, email = email)
+                } else {
+                    state.value =
+                        LoginUiState.Idle(isGenericError = true, isCodeSent = false, email = email)
+                }
+                Log.e(VIEWMODEL_ERROR_TAG, "${state.value}")
+            }
+        }
+    }
+
+    fun loginClicked(email: String, code: String) {
+        viewModelScope.launch {
+            state.value = LoginUiState.Loading(isCodeSent = true, email = email)
+            if (!isValidEmail(email) || !isValidCode(code)) {
+                state.value = LoginUiState.Idle(isEmailError = true, email = email)
+                cancel()
```

Comment:
Cancel can throw and generally is never needed unless there's a long running operation or loop in a coroutine. Just let it naturally run its course

Reply:
removed cancels